File: 2415-Reverse Odd Levels of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def reverseOddLevels(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: Optional[TreeNode]
        """
        queue = collections.deque([root])
        isOdd = False
        
        while queue:
            if isOdd:
                for i in range(len(queue)//2):
                    queue[i].val, queue[~i].val = queue[~i].val, queue[i].val
            for _ in range(len(queue)):
                curr = queue.popleft()
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
                    
            isOdd = not isOdd
        
        return root

File: C++/2415-Reverse Odd Levels of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* reverseOddLevels(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        int depth = 0;
        vector<TreeNode*> children;  // declare outside the while loop to avoid repeated memory allocation and deallocation during each iteration

        while (!q.empty()) {
            int length = q.size();
            while (length-- > 0) {
                TreeNode* curr = q.front();
                if (depth%2) children.push_back(curr);
                q.pop();
                if (curr->left != nullptr) q.push(curr->left);
                if (curr->right != nullptr) q.push(curr->right);                
            }
            int left = 0, right = children.size()-1;
            while (left < right) {
                swap(children[left]->val, children[right]->val);
                left++; right--;
            }
            children.clear();
            depth++;
        }

        return root;
    }
};